File: scripts/generate_simple_mongodb_data.js
// MongoDB data generation script
// This script generates sample data for MongoDB collections

// Sample product and customer IDs
const SAMPLE_PRODUCT_IDS = Array.from({length: 100}, (_, i) => `PROD-${String(i+1).padStart(4, '0')}`);
const SAMPLE_CUSTOMER_IDS = Array.from({length: 50}, (_, i) => `CUST-${String(i+1).padStart(4, '0')}`);

// Connect to the database
const conn = new Mongo();
const db = conn.getDB("document_store");

// Drop existing collections
db.product_details.drop();
db.customer_feedback.drop();
db.support_tickets.drop();
db.knowledge_base.drop();

// Generate product details
print("Generating product details...");
const productDetails = [];

for (let i = 0; i < 50; i++) {
    const productId = SAMPLE_PRODUCT_IDS[i];
    
    productDetails.push({
        product_id: productId,
        detailed_description: `Detailed description for product ${productId}`,
        specifications: {
            dimensions: {
                length: Math.random() * 100,
                width: Math.random() * 100,
                height: Math.random() * 100,
                unit: ["cm", "mm", "inches"][Math.floor(Math.random() * 3)]
            },
            weight: {
                value: Math.random() * 50,
                unit: ["kg", "g", "lbs"][Math.floor(Math.random() * 3)]
            },
            materials: Array.from({length: Math.floor(Math.random() * 4) + 1}, (_, j) => `Material ${j+1}`),
            colors: Array.from({length: Math.floor(Math.random() * 3) + 1}, (_, j) => `Color ${j+1}`)
        },
        features: Array.from({length: Math.floor(Math.random() * 5) + 3}, (_, j) => `Feature ${j+1}`),
        benefits: Array.from({length: Math.floor(Math.random() * 3) + 2}, (_, j) => `Benefit ${j+1}`),
        usage_instructions: `Usage instructions for product ${productId}`,
        warranty_info: {
            duration: [1, 2, 3, 5, 10][Math.floor(Math.random() * 5)],
            coverage: `Warranty coverage for product ${productId}`,
            limitations: `Warranty limitations for product ${productId}`
        },
        media: {
            images: Array.from({length: Math.floor(Math.random() * 5) + 2}, (_, j) => `product_${productId}_image_${j+1}.jpg`),
            videos: Array.from({length: Math.floor(Math.random() * 2) + 1}, (_, j) => `product_${productId}_video_${j+1}.mp4`),
            documents: Array.from({length: Math.floor(Math.random() * 2) + 1}, (_, j) => `product_${productId}_doc_${j+1}.pdf`)
        },
        related_products: SAMPLE_PRODUCT_IDS.filter(pid => pid !== productId).slice(0, Math.floor(Math.random() * 5)),
        metadata: {
            created_at: new Date(Date.now() - Math.random() * 730 * 24 * 60 * 60 * 1000).toISOString(),
            updated_at: new Date(Date.now() - Math.random() * 180 * 24 * 60 * 60 * 1000).toISOString(),
            created_by: `user${Math.floor(Math.random() * 100) + 1}@example.com`,
            version: `${Math.floor(Math.random() * 5) + 1}.${Math.floor(Math.random() * 10)}`
        }
    });
}

// Insert product details
db.product_details.insertMany(productDetails);
print(`Inserted ${productDetails.length} product details`);

// Generate customer feedback
print("Generating customer feedback...");
const customerFeedback = [];

for (let i = 0; i < 100; i++) {
    const sentiments = ["positive", "neutral", "negative"];
    const sentiment = sentiments[Math.floor(Math.random() * sentiments.length)];
    
    let rating;
    if (sentiment === "negative") {
        rating = Math.floor(Math.random() * 5) + 1;
    } else if (sentiment === "neutral") {
        rating = Math.floor(Math.random() * 2) + 3;
    } else {
        rating = Math.floor(Math.random() * 2) + 4;
    }
    
    const feedbackDate = new Date(Date.now() - Math.random() * 365 * 24 * 60 * 60 * 1000);
    
    customerFeedback.push({
        feedback_id: `FEED-${String(i+1).padStart(4, '0')}`,
        customer_id: SAMPLE_CUSTOMER_IDS[Math.floor(Math.random() * SAMPLE_CUSTOMER_IDS.length)],
        product_id: SAMPLE_PRODUCT_IDS[Math.floor(Math.random() * SAMPLE_PRODUCT_IDS.length)],
        rating: rating,
        sentiment: sentiment,
        title: `Feedback title ${i+1}`,
        content: `Feedback content ${i+1} with ${rating > 3 ? 'positive' : 'negative'} sentiment.`,
        feedback_date: feedbackDate.toISOString(),
        purchase_verified: Math.random() < 0.75,
        helpful_votes: Math.floor(Math.random() * 100),
        media: {
            images: Array.from({length: Math.floor(Math.random() * 3)}, (_, j) => `feedback_image_${j+1}.jpg`)
        },
        tags: ["quality", "price", "delivery", "service", "durability", "design", "functionality", "ease of use"]
            .sort(() => 0.5 - Math.random()).slice(0, Math.floor(Math.random() * 4)),
        response: {
            has_response: Math.random() < 0.7,
            response_date: Math.random() < 0.7 ? new Date(feedbackDate.getTime() + Math.random() * 14 * 24 * 60 * 60 * 1000).toISOString() : null,
            response_content: Math.random() < 0.7 ? `Response to feedback ${i+1}` : null,
            responder: Math.random() < 0.7 ? `Support Agent ${Math.floor(Math.random() * 10) + 1}` : null
        }
    });
}

// Insert customer feedback
db.customer_feedback.insertMany(customerFeedback);
print(`Inserted ${customerFeedback.length} customer feedback documents`);

// Generate support tickets
print("Generating support tickets...");
const supportTickets = [];

const ticketTypes = ["technical_issue", "billing_inquiry", "product_question", "return_request", "complaint"];
const priorities = ["low", "medium", "high", "critical"];
const statuses = ["open", "in_progress", "waiting_on_customer", "waiting_on_third_party", "resolved", "closed"];

for (let i = 0; i < 100; i++) {
    const createdDate = new Date(Date.now() - Math.random() * 365 * 24 * 60 * 60 * 1000);
    
    const status = statuses[Math.floor(Math.random() * statuses.length)];
    let resolution = null;
    let resolvedDate = null;
    let satisfactionRating = null;
    
    if (status === "resolved" || status === "closed") {
        resolution = {
            resolution_type: ["fixed", "workaround", "not_reproducible", "by_design", "no_fix_required"][Math.floor(Math.random() * 5)],
            resolution_note: `Resolution note for ticket ${i+1}`
        };
        resolvedDate = new Date(createdDate.getTime() + Math.random() * 30 * 24 * 60 * 60 * 1000).toISOString();
        satisfactionRating = Math.floor(Math.random() * 5) + 1;
    }
    
    supportTickets.push({
        ticket_id: `TCKT-${String(i+1).padStart(4, '0')}`,
        customer_id: SAMPLE_CUSTOMER_IDS[Math.floor(Math.random() * SAMPLE_CUSTOMER_IDS.length)],
        product_id: SAMPLE_PRODUCT_IDS[Math.floor(Math.random() * SAMPLE_PRODUCT_IDS.length)],
        subject: `Ticket subject ${i+1}`,
        description: `Ticket description ${i+1}`,
        type: ticketTypes[Math.floor(Math.random() * ticketTypes.length)],
        priority: priorities[Math.floor(Math.random() * priorities.length)],
        status: status,
        created_at: createdDate.toISOString(),
        updated_at: new Date(createdDate.getTime() + Math.random() * 30 * 24 * 60 * 60 * 1000).toISOString(),
        resolution: resolution,
        resolved_at: resolvedDate,
        satisfaction_rating: satisfactionRating,
        attachments: Array.from({length: Math.floor(Math.random() * 2)}, (_, j) => `ticket_${i+1}_attachment_${j+1}.pdf`),
        tags: ["hardware", "software", "billing", "shipping", "returns", "warranty", "damage", "missing_parts"]
            .sort(() => 0.5 - Math.random()).slice(0, Math.floor(Math.random() * 3)),
        notes: Array.from({length: Math.floor(Math.random() * 5)}, (_, j) => ({
            timestamp: new Date(createdDate.getTime() + j * 24 * 60 * 60 * 1000 + Math.random() * 8 * 60 * 60 * 1000).toISOString(),
            author: `Agent ${Math.floor(Math.random() * 10) + 1}`,
            content: `Note ${j+1} for ticket ${i+1}`,
            internal: Math.random() < 0.5
        }))
    });
}

// Insert support tickets
db.support_tickets.insertMany(supportTickets);
print(`Inserted ${supportTickets.length} support tickets`);

// Generate knowledge base articles
print("Generating knowledge base articles...");
const knowledgeBase = [];

const categories = ["troubleshooting", "how-to", "faq", "product-info", "policy"];

for (let i = 0; i < 30; i++) {
    const createdDate = new Date(Date.now() - Math.random() * 730 * 24 * 60 * 60 * 1000);
    const updatedDate = new Date(createdDate.getTime() + Math.random() * 365 * 24 * 60 * 60 * 1000);
    
    // Select random products this article applies to
    const applicableProducts = SAMPLE_PRODUCT_IDS
        .sort(() => 0.5 - Math.random())
        .slice(0, Math.floor(Math.random() * 10) + 1);
    
    knowledgeBase.push({
        article_id: `KB-${String(i+1).padStart(4, '0')}`,
        title: `Knowledge Base Article ${i+1}`,
        category: categories[Math.floor(Math.random() * categories.length)],
        content: `Content for knowledge base article ${i+1}. This is a detailed explanation.`,
        applicable_products: applicableProducts,
        tags: ["setup", "installation", "troubleshooting", "maintenance", "repair", "warranty", "returns", "usage"]
            .sort(() => 0.5 - Math.random()).slice(0, Math.floor(Math.random() * 3) + 1),
        created_at: createdDate.toISOString(),
        updated_at: updatedDate.toISOString(),
        author: `Author ${Math.floor(Math.random() * 10) + 1}`,
        view_count: Math.floor(Math.random() * 10000),
        helpful_rating: Math.floor(Math.random() * 50) / 10 + 1,
        related_articles: Array.from({length: Math.floor(Math.random() * 3)}, () => 
            `KB-${String(Math.floor(Math.random() * 30) + 1).padStart(4, '0')}`)
    });
}

// Insert knowledge base articles
db.knowledge_base.insertMany(knowledgeBase);
print(`Inserted ${knowledgeBase.length} knowledge base articles`);

// Create text indexes on all collections
print("Creating text indexes on collections...");

// Product details text index
db.product_details.createIndex(
    { 
        product_id: "text",
        detailed_description: "text",
        "specifications.materials": "text",
        "specifications.colors": "text",
        features: "text",
        benefits: "text",
        usage_instructions: "text",
        "warranty_info.coverage": "text",
        "warranty_info.limitations": "text"
    },
    { name: "product_details_text_index" }
);
print("Created text index on product_details collection");

// Customer feedback text index
db.customer_feedback.createIndex(
    {
        feedback_id: "text",
        customer_id: "text",
        product_id: "text",
        sentiment: "text",
        title: "text",
        content: "text",
        tags: "text",
        "response.response_content": "text"
    },
    { name: "customer_feedback_text_index" }
);
print("Created text index on customer_feedback collection");

// Support tickets text index
db.support_tickets.createIndex(
    {
        ticket_id: "text",
        customer_id: "text",
        product_id: "text",
        subject: "text",
        description: "text",
        type: "text",
        priority: "text",
        status: "text",
        tags: "text",
        "notes.content": "text",
        "resolution.resolution_note": "text"
    },
    { name: "support_tickets_text_index" }
);
print("Created text index on support_tickets collection");

// Knowledge base text index
db.knowledge_base.createIndex(
    {
        article_id: "text",
        title: "text",
        category: "text",
        content: "text",
        applicable_products: "text",
        tags: "text",
        author: "text"
    },
    { name: "knowledge_base_text_index" }
);
print("Created text index on knowledge_base collection");

// Create regular indexes for common query fields
db.product_details.createIndex({ product_id: 1 });
db.customer_feedback.createIndex({ product_id: 1 });
db.customer_feedback.createIndex({ customer_id: 1 });
db.support_tickets.createIndex({ product_id: 1 });
db.support_tickets.createIndex({ customer_id: 1 });
db.support_tickets.createIndex({ status: 1 });
db.knowledge_base.createIndex({ applicable_products: 1 });
db.knowledge_base.createIndex({ category: 1 });

print("MongoDB data generation and indexing completed successfully!");

File: scripts/generate_simple_mongodb_data.py
#!/usr/bin/env python
"""
Script to generate simple MongoDB data and save it to a JSON file.
This only needs to be run once to create the data file.
"""
import json
import logging
import os
import random
import sys
from datetime import datetime, timedelta

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Sample product and customer IDs
SAMPLE_PRODUCT_IDS = [f"PROD-{i:04d}" for i in range(1, 101)]
SAMPLE_CUSTOMER_IDS = [f"CUST-{i:04d}" for i in range(1, 51)]

def generate_product_details(num_products=50, seed=42):
    """Generate product details."""
    random.seed(seed)
    
    documents = []
    selected_product_ids = random.sample(SAMPLE_PRODUCT_IDS, min(num_products, len(SAMPLE_PRODUCT_IDS)))
    
    for product_id in selected_product_ids:
        document = {
            "product_id": product_id,
            "detailed_description": f"Detailed description for product {product_id}",
            "specifications": {
                "dimensions": {
                    "length": round(random.uniform(1, 100), 2),
                    "width": round(random.uniform(1, 100), 2),
                    "height": round(random.uniform(1, 100), 2),
                    "unit": random.choice(["cm", "mm", "inches"])
                },
                "weight": {
                    "value": round(random.uniform(0.1, 50), 2),
                    "unit": random.choice(["kg", "g", "lbs"])
                },
                "materials": [f"Material {i}" for i in range(1, random.randint(2, 6))],
                "colors": [f"Color {i}" for i in range(1, random.randint(2, 5))]
            },
            "features": [f"Feature {i}" for i in range(1, random.randint(4, 9))],
            "benefits": [f"Benefit {i}" for i in range(1, random.randint(3, 6))],
            "usage_instructions": f"Usage instructions for product {product_id}",
            "warranty_info": {
                "duration": random.choice([1, 2, 3, 5, 10]),
                "coverage": f"Warranty coverage for product {product_id}",
                "limitations": f"Warranty limitations for product {product_id}"
            },
            "media": {
                "images": [f"product_{product_id}_image_{i}.jpg" for i in range(1, random.randint(3, 8))],
                "videos": [f"product_{product_id}_video_{i}.mp4" for i in range(1, random.randint(1, 3))],
                "documents": [f"product_{product_id}_doc_{i}.pdf" for i in range(1, random.randint(1, 3))]
            },
            "related_products": random.sample(
                [pid for pid in SAMPLE_PRODUCT_IDS if pid != product_id],
                min(random.randint(0, 5), len(SAMPLE_PRODUCT_IDS) - 1)
            ),
            "metadata": {
                "created_at": (datetime.now() - timedelta(days=random.randint(0, 730))).isoformat(),
                "updated_at": (datetime.now() - timedelta(days=random.randint(0, 180))).isoformat(),
                "created_by": f"user{random.randint(1, 100)}@example.com",
                "version": f"{random.randint(1, 5)}.{random.randint(0, 9)}"
            }
        }
        documents.append(document)
    
    return documents

def generate_customer_feedback(num_feedback=100, seed=43):
    """Generate customer feedback."""
    random.seed(seed)
    
    documents = []
    
    for i in range(num_feedback):
        sentiment = random.choice(["positive", "neutral", "negative"])
        rating = random.randint(1, 5) if sentiment == "negative" else \
                 random.randint(3, 4) if sentiment == "neutral" else \
                 random.randint(4, 5)
        
        feedback_date = datetime.now() - timedelta(days=random.randint(0, 365))
        
        document = {
            "feedback_id": f"FEED-{i:04d}",
            "customer_id": random.choice(SAMPLE_CUSTOMER_IDS),
            "product_id": random.choice(SAMPLE_PRODUCT_IDS),
            "rating": rating,
            "sentiment": sentiment,
            "title": f"Feedback title {i}",
            "content": f"Feedback content {i} with {'positive' if rating > 3 else 'negative'} sentiment.",
            "feedback_date": feedback_date.isoformat(),
            "purchase_verified": random.choice([True, False, True, True]),
            "helpful_votes": random.randint(0, 100),
            "media": {
                "images": [f"feedback_image_{j}.jpg" for j in range(random.randint(0, 3))]
            },
            "tags": random.sample(
                ["quality", "price", "delivery", "service", "durability", "design", "functionality", "ease of use"],
                random.randint(0, 4)
            ),
            "response": {
                "has_response": random.choice([True, False]),
                "response_date": (feedback_date + timedelta(days=random.randint(1, 14))).isoformat() 
                if random.random() < 0.7 else None,
                "response_content": f"Response to feedback {i}" if random.random() < 0.7 else None,
                "responder": f"Support Agent {random.randint(1, 10)}" if random.random() < 0.7 else None
            }
        }
        documents.append(document)
    
    return documents

def generate_support_tickets(num_tickets=100, seed=44):
    """Generate support tickets."""
    random.seed(seed)
    
    documents = []
    ticket_types = ["technical_issue", "billing_inquiry", "product_question", "return_request", "complaint"]
    priorities = ["low", "medium", "high", "critical"]
    statuses = ["open", "in_progress", "waiting_on_customer", "waiting_on_third_party", "resolved", "closed"]
    
    for i in range(num_tickets):
        created_date = datetime.now() - timedelta(days=random.randint(0, 365))
        
        status = random.choice(statuses)
        resolution = None
        resolved_date = None
        satisfaction_rating = None
        
        if status in ["resolved", "closed"]:
            resolution = {
                "resolution_type": random.choice(["fixed", "workaround", "not_reproducible", "by_design", "no_fix_required"]),
                "resolution_note": f"Resolution note for ticket {i}"
            }
            resolved_date = (created_date + timedelta(days=random.randint(1, 30))).isoformat()
            satisfaction_rating = random.randint(1, 5)
        
        document = {
            "ticket_id": f"TCKT-{i:04d}",
            "customer_id": random.choice(SAMPLE_CUSTOMER_IDS),
            "product_id": random.choice(SAMPLE_PRODUCT_IDS),
            "subject": f"Ticket subject {i}",
            "description": f"Ticket description {i}",
            "type": random.choice(ticket_types),
            "priority": random.choice(priorities),
            "status": status,
            "created_at": created_date.isoformat(),
            "updated_at": (created_date + timedelta(days=random.randint(0, 30))).isoformat(),
            "resolution": resolution,
            "resolved_at": resolved_date,
            "satisfaction_rating": satisfaction_rating,
            "attachments": [f"ticket_{i}_attachment_{j}.pdf" for j in range(random.randint(0, 2))],
            "tags": random.sample(
                ["hardware", "software", "billing", "shipping", "returns", "warranty", "damage", "missing_parts"],
                random.randint(0, 3)
            ),
            "notes": [
                {
                    "timestamp": (created_date + timedelta(days=j, hours=random.randint(1, 8))).isoformat(),
                    "author": f"Agent {random.randint(1, 10)}",
                    "content": f"Note {j} for ticket {i}",
                    "internal": random.choice([True, False])
                }
                for j in range(random.randint(0, 5))
            ]
        }
        documents.append(document)
    
    return documents

def generate_knowledge_base(num_articles=30, seed=45):
    """Generate knowledge base articles."""
    random.seed(seed)
    
    documents = []
    categories = ["troubleshooting", "how-to", "faq", "product-info", "policy"]
    
    for i in range(num_articles):
        created_date = datetime.now() - timedelta(days=random.randint(0, 730))
        updated_date = created_date + timedelta(days=random.randint(0, 365))
        
        # Select random products this article applies to
        applicable_products = random.sample(
            SAMPLE_PRODUCT_IDS,
            random.randint(1, 10)
        )
        
        document = {
            "article_id": f"KB-{i:04d}",
            "title": f"Knowledge Base Article {i}",
            "category": random.choice(categories),
            "content": f"Content for knowledge base article {i}. This is a detailed explanation.",
            "applicable_products": applicable_products,
            "tags": random.sample(
                ["setup", "installation", "troubleshooting", "maintenance", "repair", "warranty", "returns", "usage"],
                random.randint(1, 4)
            ),
            "created_at": created_date.isoformat(),
            "updated_at": updated_date.isoformat(),
            "author": f"Author {random.randint(1, 10)}",
            "view_count": random.randint(0, 10000),
            "helpful_rating": round(random.uniform(1, 5), 1),
            "related_articles": [f"KB-{random.randint(1, num_articles):04d}" for _ in range(random.randint(0, 3))]
        }
        documents.append(document)
    
    return documents

def main():
    """Generate MongoDB data and save it to a JSON file."""
    output_file = "mongodb_data.json"
    if len(sys.argv) > 1:
        output_file = sys.argv[1]
    
    logger.info(f"Generating MongoDB data to {output_file}...")
    
    # Generate data for each collection
    collections_data = {
        "product_details": generate_product_details(),
        "customer_feedback": generate_customer_feedback(),
        "support_tickets": generate_support_tickets(),
        "knowledge_base": generate_knowledge_base()
    }
    
    # Save the data to a JSON file
    with open(output_file, 'w') as f:
        json.dump(collections_data, f, indent=2)
    
    logger.info(f"MongoDB data saved to {output_file}")
    logger.info(f"Generated {len(collections_data)} collections with a total of {sum(len(docs) for docs in collections_data.values())} documents")

if __name__ == "__main__":
    main()